pkg/ingester/active_series.go

```diff
@@ -25,11 +25,10 @@ const activeSeriesMaxSizeBytes = 1 * 1024 * 1024
 // ActiveSeries implements the ActiveSeries RPC. It returns a stream of active
 // series that match the given matchers.
 func (i *Ingester) ActiveSeries(request *client.ActiveSeriesRequest, stream client.Ingester_ActiveSeriesServer) (err error) {
-	defer func() { err = i.mapReadErrorToErrorWithStatus(err) }()
-	if err := i.checkAvailableForRead(); err != nil {
-		return err
-	}
-	if err := i.checkReadOverloaded(); err != nil {
+	var st *readRequestState
+	defer func() { err = i.finishReadRequest(st, err) }()
+	st, err = i.startReadRequest()
+	if err != nil {
 		return err
 	}
```

Comment:
Shouldn't this be like:
```suggestion
	var st *readRequestState
	st, err = i.startReadRequest()
	if err != nil {
		return err
	}
	defer func() { err = i.finishReadRequest(st, err) }()
```

We shouldn't defer the finishReadRequest if there was an error, right?

Reply:
Well not really... In the previous implementation we had 
```
defer func() { err = i.mapReadErrorToErrorWithStatus(err) }
```
before we called the starting checks. This is because all auxiliary methods return errors of type `ingesterError`, which are then converted to gRPC errors by calling ` i.mapReadErrorToErrorWithStatus()`. That's why the latter had to be executed in any scenario, and therefore the defer had to be placed at the top of the code.
The logic here is similar. The only difference is that if `startReadRequest()` (doesn't) acquires a cb permit, the `finishReadRequest()` will (not) release the permit, and whether the permit was acquired or not is stored in the `readRequestState` object returned by `startReadRequest()`

Reply:
We might have this scenario:
- we want to execute `QueryStream()`, so we first execute `startReadRequest()`
- `startReadRequest()` acquires a cb permit
- `startReadRequest()` checks if the ingester is overloaded and it realizes it is, meaning that we are not going to start the push request, and we return `errTooBusy`
- `QueryStream()` should now fail. If we don't call `finishPushRequest()` we will not release the permit

Reply:
I find it very confusing that we call `defer finish...()` before caling the thing that gives you the thing you're finishing. That's something we usually don't see in golang, and IMO it's a smell.

Can we split this into separate calls and defers, making sure that we only defer stuff that has been acquired already?